hashicorp/consul-k8s
# Use kube auth method to provision ACL token for the crd controller

Changes proposed in this PR:
- Adds a new auth method called `release-name-consul-k8s-component-auth-method` which will be used for components to issue `ACL().Login()` at startup. 
- Modify the `acl-init` init container to issue an `ACL().Login()` to fetch the component's ACL token on startup and store it locally, using the aforementioned new authMethod.
- Modifies `server-acl-init` job to create the new auth method and also no longer provision Kube secrets for the controller.
- Modifies the `server-acl-init` unit tests to support cases where we do not have a 1:1 mapping of policies:kube tokens. 
- Adds a new subcommand `consul-logout` which is used as a `preStop` hook for component pods to destroy the ACL token that was created through `acl-init`.
- move the logic of `Login()` that resides in connect-init/command.go to common/ and use this in both acl-init and connect-init.
- Applies the above features to the CRD controller deployment.
- Adds bats tests coverage.
- always deploy the connect-inject serviceaccount + role + rolebinding, this bhehaviour will be changed in a follow-on PR

How I've tested this PR:
* Manual testing using the below steps
* Update unit tests
* Add unit tests

How I expect reviewers to test this PR:
Check out the associated branch and install using the following val.yaml file:
```
demo $ cat val.yaml
global:
  image: hashicorp/consul:1.11.3
  imageK8S: kyleschochenmaier/consul-k8s-acls
  tls:
    enabled: true
  acls:
    manageSystemACLs: true
controller:
  enabled: true
connectInject:
  enabled: true
ui:
  enabled: true
server:
  enabled: true
  replicas: 1
```
* Install using helm:  `helm install kyle-consul ~/go/src/github.com/hashicorp/consul-k8s/charts/consul -f val.yaml`
* Once everything is online you can check logs :  `demo $ k logs -f deployment/kyle-consul-consul-controller -c controller-acl-init`  and look for the ACL login at the end.
* `kubectl get secrets` will confirm that we've not created an (opaque) ACL token for the controller.
* open the UI and navigate to the ACL tokens window, then delete the controller pod and see the old ACL token cleaned up and a new one provisioned when the pod starts.

Checklist:
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)
